enhancements/microshift/multus-cni-for-microshift.md

```diff
@@ -0,0 +1,445 @@
+relevant parts of network set up.
+This should not be an issue because Pods that do not use `hostNetwork` will have networking setup
+after the CNI is ready. Both ovn-kubernetes and Multus use `hostNetwork` so they start before other
+Pods. If CRI-O is configured to use the Multus regardless of files in /etc/cni/net.d, CRI-O will wait
+for the Multus. We can also use the Multus' config option of `--readiness-indicator-file` to make
+sure the Multus waits for the ovn-kubernetes.
+
+### Drawbacks
+
+This section includes limitations of the Multus itself, not its integration with MicroShift.
+However, these drawbacks should be documented nonetheless.
+
+Multus does not actively watch NetworkAttachmentDefinitions or annotations therefore to make changes
+to these resources effective the Pod must be re-created. This behavior is reasonable because it does not
+disrupt the Pod's networking and should not be a problem in production environments where we expect to
+have stable configurations.
+
+Multus also does not observe the underlying bridge interfaces, therefore if one is rebuilt, the Pod's
+interface might stop working (see [BZ #2066351](https://bugzilla.redhat.com/show_bug.cgi?id=2066351)).
+If these limitations are ever addressed (see [NP-606](https://issues.redhat.com/browse/NP-606) and 
+[NP-608](https://issues.redhat.com/browse/NP-608)), they would, most likely, be part of the thick
+Multus plugin.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Clean up of Multus artifacts on disk during system rollback or when `microshift-cleanup-data` is executed:
+   Should we invest in creating a lightweight plugin architecture to avoid adding cleanup of Multus
+   to MicroShift's core cleanup script? This architecture could be based on existing
+   [kubectl plugin](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) architecture
+   where plugins start with `kubectl-`, so we could use `microshift-cleanup-plugin-` prefix.
+
+### Test Plan
+
+Multus CNI for MicroShift will be tested using existing test harness. New test suite will be created
+with simple test for each CNI we want to declare as supported.
+
+Starting with `bridge` test shall: install Multus, create NetworkAttachmentDefinition and deploy
+a Pod that should have additional interface attached. After verifying that the interface is present,
+test will try to access Pod's application using the bridge interface on the host to make sure there
+is a connectivity.
+
+Tests for other networking plugins will be designed and implemented when plugins are planned for support.
+
+### Graduation Criteria
+
+Multus CNI for MicroShift is targeted to be GA next release.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+Because both Multus and MicroShift RPMs will be built from the same spec file, they will share the
```

Comment:
Do we consider upgrading from a 4.15 cluster to a multus-enabled 4.16 cluster? or is it not possible.
There might be some cleanup work needs to be done for the upgraded cluster to use Multus correctly (for example, remove the old node CNI config file written by ovn-kubernetes, since the new config file will be created in another directory)

Reply:
I actually decided to not change the directories: keep configs in `/etc/cni/net.d` and binaries in `/opt/cni/bin` so it's easy plug n play.

Reply:
If we keep the config in the same directory, would there be such situation that:
1. ovnk writes its CNI config under /etc/cni/net.d
2. node become ready
3. pod starts with ovnk CNI
4. multus pod ready and update the CNI config under /etc/cni/net.d
5. other pods start with Multus CNI

Is it guaranteed that multus pod becomes ready before ovnk?

Reply:
never mind, I learned in the previous comment that we want to force a reboot after multus installation which shall guarantee that crio use the multus generated CNI config file.

Reply:
Yes. Also, from my testing, if we configure CRI-O with `cni_default_network = "multus"`, then even if the ovn-k's config exists, the Node will not be ready because of "no ready cni"

Reply:
How much does enabling Multus change the startup timing of the other core MicroShift pods? Does the healthcheck used by the greenboot script need to be updated to take that into account, or are we still within the limits built into the script?

Reply:
I don't have data to back this up, but creating cluster with images present I expect startup of multus to be quicker than ovnk's (Multus will wait for ovnk).